feat(build-tools): policy for entrypoint linting

jason-ha opened the pull request:
Policy:
- Package "exports" other than those under "./internal" should be linted by api-extractor.
- A single "./internal" or alternatively "." export should also be linted by api-extractor with a bundle config. Recently this was "check:release-tags" and now is "check:exports:bundle-release-tags".

Unfortunately, api-extractor will only lint a single entrypoint at a time and needs the config file to be local. So, this means lots of files and script entries.

Build config:
 - Add "check:exports" to "lint" tasks.

Apply policy to client-utils and container-definitions.

[AB#8141](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/8141)

jason-ha committed:
feat(build-tools): policy for entrypoint linting

package "exports" other than those under internal should be linted by api-extractor.

A single /internal export should also be linted by api-extractor. Recently this was check:release-tags and now is check:exports:bundle-release-tags.

Add check:exports to build tasks.

jason-ha committed:
test(client): apply API lint policy
to client-utils and container-definitions

tylerbutler commented on build-tools/packages/build-cli/src/library/repoPolicyCheck/npmPackages.ts:
```diff
@@ -1443,6 +1652,96 @@ export const handlers: Handler[] = [
+				)} should be linted by check:exports:* script using api-extractor.`;
+			}
+		},
+		resolver: async (
```
Asynchronous resolvers... how do you ensure that a resolver doesn't overwrite changes by other asynchronous resolvers?

jason-ha replied:
The tool scheduling the resolvers waits for each one. That already appeared to be the case.

jason-ha replied:
I think we might be in middle ground. routeToHandlers already has this code:
```typescript
						const { resolver } = handler;
						if (this.flags.fix && resolver) {
							output += `${newline}attempting to resolve: ${relPath}`;
							const resolveResult = await runWithPerf(handler.name, "resolve", async () =>
								resolver(relPath, gitRoot),
							);
```
that I didn't have to fix up.

But that doesn't really mean that everything is well scheduled. Looking at the surrounding code it says, if there is a handler that is async, then resolvers will run on arbitrary schedules.

jason-ha replied:
I split the async support to #21249

tylerbutler commented on build-tools/packages/build-cli/src/library/repoPolicyCheck/npmPackages.ts:
```diff
@@ -505,6 +509,211 @@ function getPreferredScriptLine(scriptLine: string): string {
+
+	// Make sure `concurrently` is available.
+	if (packageJson.devDependencies?.concurrently === undefined) {
+		devDependencies.push("concurrently");
```
There's already a policy that checks that dependencies exist for commands used in the package scripts. Why is that policy not catching this?

jason-ha replied:
It very well could (what is the policy?), but if this introducing the dependency, then it should fix it, I think, rather than requiring a second pass of fixing.

tylerbutler replied:
> if this introducing the dependency, then it should fix it, I think, rather than requiring a second pass of fixing.

Good point. The policy I'm referring to is called npm-package-json-script-dep and is configured here. Looks like concurrently isn't there but it should be. We ca fix that separately.

https://github.com/microsoft/FluidFramework/blob/c6293d4b833da067c9abbd447eb9483fc8425be9/fluidBuild.config.cjs#L413-L436

tylerbutler replied:
PR to fix the concurrently omission: https://github.com/microsoft/FluidFramework/pull/21259

jason-ha replied:
Unfortunately npm-package-json-script-dep doesn't have a resolver. I guess it can't pick a version. I think there is something else that cares about consistency. Maybe those two should get together to have a resolver.

jason-ha replied:
PR to add concurrently support to fluid-build: #21262

tylerbutler commented on build-tools/packages/build-cli/src/library/repoPolicyCheck/npmPackages.ts:
```diff
@@ -505,6 +509,211 @@ function getPreferredScriptLine(scriptLine: string): string {
+	// There are files that need linting.
+
+	// Make sure the package.json has the check:exports script that runs others.
+	const checkExports = packageJson.scripts?.["check:exports"];
```
There should be a policy already that lets us configure required scripts. I'd rather make several smaller policies that check only one or two things and combine them rather than one policy that checks a lot of different things.

jason-ha replied:
I'll go with the collective wisdom.
I view this all as one policy that exports are linted. There are multiple pieces to that policy. I don't like it if there is one thing that has to be coordinated in several places without a clear thing tying them together.

tylerbutler reviewed (commented):
First impressions

Wow that's a lot of config files. Yuck.

The policy seems very complex. In my experience complex policies are very hard to maintain and end up being brittle and less useful than policies that are 'dumb'. Some of the checks in this policy should already be handled by other policies, or could be split out as configurable individual policies.

Async resolvers seem like might cause issues with ordering/concurrently writing files etc.

tylerbutler commented on build-tools/packages/build-cli/src/library/repoPolicyCheck/common.ts:
```diff
@@ -14,7 +14,12 @@ export interface Handler {
+		file: string,
+		root: string,
+	) =>
+		| Promise<{ resolved: boolean; message?: string }>
+		| { resolved: boolean; message?: string };
```
Any reason the resolvers shouldn't always be async?

jason-ha replied:
They can be, but there isn't a reason to force them to be async. If a resolver isn't returning a Promise, then requiring Promise causes build breaks. TFoo does not satisfy Promise<TFoo>. If we do want them all async, then that should be a follow-up.

Josmithr commented:
> Wow that's a lot of config files. Yuck.

Plus 1 to this. This seems untenable to me.

API-Extractor does have a programmatic API. I wonder if it might be worth exploring options using that, rather than their CLI. We could generate config files dynamically that way based on some simpler config setup, and invoke API-Extractor programmatically in a loop within a single npm script (or something).

I'm also worried about the performance here. Have you evaluated the cost of this yet?

jason-ha commented:
> > Wow that's a lot of config files. Yuck.
> 
> Plus 1 to this. This seems untenable to me.
> 
> API-Extractor does have a programmatic API. I wonder if it might be worth exploring options using that, rather than their CLI. We could generate config files dynamically that way based on some simpler config setup, and invoke API-Extractor programmatically in a loop within a single npm script (or something).
> 
> I'm also worried about the performance here. Have you evaluated the cost of this yet?

I think we all agree. I think the right solution is to get new feature(s) from api-extractor.
I wrote a policy with fixer because it is so bumpy. The policy is pretty simple, we need to lint all of our exports.
We are not in favor of writing our own tools if we can help it and it gets complicated using api-extractor programmatically. So, I think this is the best current option. And we can replace it in the future. Well, a lot of the policy logic remains after, but the noise in package.json files will be reduced.
If we want less noise; we could limit the checks to ESM and we should feel pretty confident, but as we go to 2.0 I would rather have broad coverage that we roll back even if it hits build time.

We need this effective policy if we are going to reclaim `@alpha` tag. And I am hoping we can.
cc @Josmithr , @tylerbutler

Josmithr commented:
> We need this effective policy if we are going to reclaim @alpha tag. And I am hoping we can.

I'm not sure I understand this. Or do you mean, we need this before we can start using manually crafted entrypoints in addition to release-tag-derived ones?

> We are not in favor of writing our own tools if we can help it and it gets complicated using api-extractor programmatically. 

I agree that we should avoid writing our own tooling where we can help it, but it seems like writing a small wrapper around the programmatic API to invoke API-Extractor with a few different `d.ts` file targets shouldn't be too difficult. Likely a much quicker solution than rolling out new functionality in API-Extractor, and probably better IMO than adding all of these config files.

> I think the right solution is to get new feature(s) from api-extractor.

Just to make sure we're on the same page, you mean:
1. Get them to support nominal types better so we can replace our custom roll-up logic (https://github.com/microsoft/rushstack/issues/4654)
2. Probably add support for multiple entrypoint files (possibly reading what's listed in `exports`)?
    - I don't know of any existing item tracking this; we should probably file something.
4. Anything else I'm missing?

jason-ha commented:
> > We need this effective policy if we are going to reclaim `@alpha` tag. And I am hoping we can.
> 
> I'm not sure I understand this. Or do you mean, we need this before we can start using manually crafted entrypoints in addition to release-tag-derived ones?

We already have some manually crafted that are not covered. Anything api-extractor doesn't natively generate would need this. I am assuming that if natively generated and linting is enabled during the generation, then the results will be good, but it is a risk.

> > We are not in favor of writing our own tools if we can help it and it gets complicated using api-extractor programmatically.
> 
> I agree that we should avoid writing our own tooling where we can help it, but it seems like writing a small wrapper around the programmatic API to invoke API-Extractor with a few different `d.ts` file targets shouldn't be too difficult. Likely a much quicker solution than rolling out new functionality in API-Extractor, and probably better IMO than adding all of these config files.

I did consider it, but it didn't have the bandwidth and didn't want to deal with some of the little hurdles. Value for the effort *now* didn't add up for me. No argument that this looks ugly, but I didn't see an immediate maintenance burden.

> > I think the right solution is to get new feature(s) from api-extractor.
> 
> Just to make sure we're on the same page, you mean:
> 
> 1. Get them to support nominal types better so we can replace our custom roll-up logic ([[api-extractor] Better support for nominal types rushstack#4654](https://github.com/microsoft/rushstack/issues/4654))
> 2. Probably add support for multiple entrypoint files (possibly reading what's listed in `exports`)?
>    
>    * I don't know of any existing item tracking this; we should probably file something.
> 3. Anything else I'm missing?

And 1 may get more complicated.
2 is what should be filed. I was also thinking that exports walk was good, but there could probably be some filtering option to skip some.

tylerbutler commented on fluidBuild.config.cjs:
```diff
@@ -47,6 +47,7 @@ module.exports = {
 				"eslint",
 				"good-fences",
 				"depcruise",
+				"check:exports",
```
You may also want to add a task definition for this, something like `"check:exports": []`. Technically that is only needed if the task is script:false, but we generally define them. Not a huge deal and happy to talk about changing the "pattern" here and only defining tasks that must be explicitly defined.

jason-ha replied:
It should have been there. And it should depend on "api" since those are the things it checks.

tylerbutler reviewed (approved):
Approving without full review to keep things moving forward.

jason-ha committed:
build(client): work around fluid-build concurrently limitation
by using explicit script names

jason-ha committed:
improvement(build-tools): keep api-extractor configs
in api-extractor folder

jason-ha committed:
build(client): add concurrently dep per policy check

jason-ha committed:
fix(build-tools): relative path support
+ use <projectFolder> in api-extractor config path to make it less location sensitive.

jason-ha committed:
fix(build-tools): make api-extractor dir during fix
+ use a root file for bundle check if ./internal isn't present to be consistent with prior general "release-tags" check
+ add @api-extractor as an expected devDep
+ fix silent failure when only non-file specific defects are found
+ add new policy as excluded to start

jason-ha committed:
test(client): exclude known problem packages from policy
